tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
 
         self::assertNotContains('<div', $subject->render());
     }
+
+    /**
+     * @test
+     */
+    public function removeRedundantClassesProvidesFluentInterface()
+    {
+        $subject = HtmlPruner::fromHtml('<html></html>');
+
+        $result = $subject->removeRedundantClasses([]);
+
+        self::assertSame($subject, $result);
+    }
+
+    /**
+     * @return string[][][]
+     */
+    public function classesToKeepDataProvider()
+    {
+        return [
+            'no classes to keep' => [[]],
+            '1 class to keep' => [['foo']],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string[] $classesToKeep
+     *
+     * @dataProvider classesToKeepDataProvider
+     */
+    public function removeRedundantClassesPreservesHtmlWithoutClasses(array $classesToKeep)
+    {
+        $html = '<p style="color: green;">hello</p>';
+        $subject = HtmlPruner::fromHtml('<html>' . $html . '</html>');
+
+        $subject->removeRedundantClasses($classesToKeep);
+
+        self::assertContains($html, $subject->render());
+    }
+
+    /**
+     * @return (string|string[])[][]
+     */
+    public function nonMatchedClassesDataProvider()
+    {
+        return [
+            '1 attribute, 1 class, no classes to keep' => [
+                'html' => '<p class="foo">hello</p>',
+                'classes to keep' => [],
+                'classes expected to be removed' => ['foo'],
```

Comment:
I like having these descriptive sub-array keys.

Reply:
I thought it might be possible, and would help readability, particularly with two arrays of classes.  So I tried it and found it was possible 😎

Though as they are for human consumption only, `'html'` should really be `'HTML'`.  I'll change that...